### ISSUE TITLE
Trainable parameter, no weight decay

### ISSUE THREAD

**Issue by homerjed** [danielward27/paramax#6]
Hi! 

Alongside `flowjax` - another great package, thanks!

I was wondering about a `paramax` wrapper for weight decay (I remember a github issue about this on `equinox` I think...).

Specifically, I would like to optimise a parameter but _not_ apply weight decay to it (so `NonTrainable` I think is not applicable here). 

Do you have any pointers for this? If not I will have a go myself!

Cheers

**Comment by homerjed** [comment]
This isn't really a paramax question per se so I think I'll close. Here is a rough but principled solution for some common moduels if anyone else needs it. 

```python

import jax
import jax.numpy as jnp
import jax.random as jr
import equinox as eqx
import optax


class ToyModel(eqx.Module):
    emb: eqx.nn.Embedding
    ln: eqx.nn.LayerNorm
    fc1: eqx.nn.Linear
    fc2: eqx.nn.Linear

    def __init__(self, *, vocab_size=100, dim=32, hidden=64, key):
        k1, k2, k3 = jr.split(key, 3)
        self.emb = eqx.nn.Embedding(vocab_size, dim, key=k1)
        self.ln = eqx.nn.LayerNorm(dim)
        self.fc1 = eqx.nn.Linear(dim, 100, key=k2)
        self.fc2 = eqx.nn.Linear(100, 1, key=k3)

    def __call__(self, token_ids: jnp.ndarray):
        x = jax.vmap(self.emb)(token_ids)     
        x = jax.vmap(self.ln)(x)              
        x = self.fc1(jnp.sum(x, axis=0))
        x = jax.nn.gelu(x)
        x = self.fc2(x)
        return x


key = jr.key(0)
model = ToyModel(key=key)


filter_spec = jax.tree.map(lambda _: True, model)

def mask_layernorm_and_embeddings(model):

    mask = jax.tree.map(lambda _: True, model)

    def is_embedding(m):
        return isinstance(m, eqx.nn.Embedding)

    def is_layernorm(m):
        return isinstance(m, eqx.nn.LayerNorm)

    # Weight of embeddings
    get_embeddings = lambda m: [
        x.weight
        for x in jax.tree.leaves(m, is_leaf=is_embedding)
        if is_embedding(x)
    ]
    mask = eqx.tree_at(
        get_embeddings, mask, replace_fn=lambda l: False 
    )

    # Weight / bias of layernorms
    get_layernorms = lambda m: [
        x.weight
        for x in jax.tree.leaves(m, is_leaf=is_layernorm)
        if is_layernorm(x)
    ]
    mask = eqx.tree_at(
        get_layernorms, mask, replace_fn=lambda l: False
    )
    get_layernorms = lambda m: [
        x.bias
        for x in jax.tree.leaves(m, is_leaf=is_layernorm)
        if is_layernorm(x)
    ]
    mask = eqx.tree_at(
        get_layernorms, mask, replace_fn=lambda l: False
    )
    return mask

mask = mask_layernorm_and_embeddings(model)

optim1 = optax.adamw(1., mask=mask, weight_decay=1.)
optim2 = optax.adamw(1., mask=mask, weight_decay=0.)

state1 =  optim1.init(eqx.filter(model, eqx.is_array))
state2 =  optim1.init(eqx.filter(model, eqx.is_array))

@eqx.filter_value_and_grad
def grad_loss(model, input, output):
    pred = model(input)
    print(pred.shape, output.shape)
    mse = lambda x, y : jnp.mean(jnp.square(x-y))
    return mse(pred, output)

@eqx.filter_jit
def make_step(input, output, model, states):
    loss, grads = grad_loss(model, input, output)
    u1, opt_state = optim1.update(grads, states[0], model)
    u2, opt_state = optim2.update(grads, states[1], model)
    is_working = (
        # These shouldn't change...
        jnp.allclose(u1.emb.weight, u2.emb.weight) 
        & jnp.allclose(u1.ln.weight, u2.ln.weight) 
        & jnp.allclose(u1.ln.bias, u2.ln.bias) 
        # ...but these can change
        & (~jnp.allclose(u1.fc1.weight, u2.fc1.weight))
        & (~jnp.allclose(u1.fc2.weight, u2.fc2.weight))
        & (~jnp.allclose(u1.fc1.bias, u2.fc1.bias))
        & (~jnp.allclose(u1.fc2.bias, u2.fc2.bias))
    )
    jax.debug.print('Did mask work?: {}', is_working)
    model = eqx.apply_updates(model, u1)
    return loss, model, opt_state

xs = jnp.ones((100,)).astype(jnp.int32)
ys = jr.normal(key, (1,))

loss, model, opt_state = make_step(xs, ys, model, (state1, state2))
```
